node/node.go

```diff
@@ -626,100 +624,120 @@ func (n *Node) waitRole(ctx context.Context, role string) error {
+	defer waitRoleCancel()
+	go func() {
+		n.waitRole(waitRoleCtx, ca.WorkerRole)
+		close(workerRole)
+	}()
 
+	defer func() {
 		n.Lock()
-		n.manager = m
+		n.manager = nil
 		n.Unlock()
+		m.Stop(ctx)
+		<-done
+		n.setControlSocket(nil)
+	}()
 
-		connCtx, connCancel := context.WithCancel(ctx)
-		go n.initManagerConnection(connCtx, ready)
-
-		// this happens only on initial start
-		if ready != nil {
-			go func(ready chan struct{}) {
-				select {
-				case <-ready:
-					addr, err := n.RemoteAPIAddr()
-					if err != nil {
-						log.G(ctx).WithError(err).Errorf("get remote api addr")
-					} else {
-						n.remotes.Observe(api.Peer{NodeID: n.NodeID(), Addr: addr}, remotes.DefaultObservationWeight)
-					}
-				case <-connCtx.Done():
-				}
-			}(ready)
-			ready = nil
-		}
+	n.Lock()
+	n.manager = m
+	n.Unlock()
 
-		roleChanged := make(chan error)
-		waitCtx, waitCancel := context.WithCancel(ctx)
-		go func() {
-			err := n.waitRole(waitCtx, ca.WorkerRole)
-			roleChanged <- err
-		}()
+	connCtx, connCancel := context.WithCancel(ctx)
+	defer connCancel()
 
-		select {
-		case <-done:
-			// Fail out if m.Run() returns error, otherwise wait for
-			// role change.
-			if runErr != nil {
-				err = runErr
-			} else {
-				err = <-roleChanged
-			}
-		case err = <-roleChanged:
-		}
+	go n.initManagerConnection(connCtx, ready)
 
-		n.Lock()
-		n.manager = nil
-		n.Unlock()
+	// this happens only on initial start
+	if ready != nil {
+		go func(ready chan struct{}) {
+			select {
+			case <-ready:
+				addr, err := n.RemoteAPIAddr()
+				if err != nil {
+					log.G(ctx).WithError(err).Errorf("get remote api addr")
+				} else {
+					n.remotes.Observe(api.Peer{NodeID: n.NodeID(), Addr: addr}, remotes.DefaultObservationWeight)
+				}
+			case <-connCtx.Done():
+			}
+		}(ready)
+	}
 
+	// wait for manager stop or for role change
+	// if manager stopped before role change, wait for new role for 16 seconds,
+	// then just restart manager, we might just miss that event.
+	// we need to wait for role to prevent manager to start again with wrong
+	// certificate
+	select {
+	case <-done:
+		timer := time.NewTimer(16 * time.Second)
+		defer timer.Stop()
+		select {
+		case <-timer.C:
+			log.G(ctx).Warn("failed to get worker role after manager stop, restart manager")
+			return runErr
+		case <-workerRole:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	case <-workerRole:
+		log.G(ctx).Info("role changed to worker, wait for manager to stop")
 		select {
 		case <-done:
 		case <-ctx.Done():
-			err = ctx.Err()
-			m.Stop(context.Background())
-			<-done
+			return ctx.Err()
 		}
-		connCancel()
-		n.setControlSocket(nil)
-		waitCancel()
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return runErr
+}
 
-		if err != nil {
+func (n *Node) superviseManager(ctx context.Context, securityConfig *ca.SecurityConfig, ready chan struct{}) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
```

Comment:
Why is this select necessary? It looks like `waitRole` will handle the context being cancelled.